Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -0,0 +1,509 @@
+                    } else {
+                        leased -= 1
+                    }
+                }
+            }
+        }
+
+        func closeAllConnections() -> EventLoopFuture<Void> {
+            return self.lock.withLock {
+                let closeFutures = availableConnections.map { $0.channel.close().recover { _ in } }
+                return EventLoopFuture<Void>.andAllSucceed(closeFutures, on: loopGroup.next()).map {
+                    self.availableConnections.removeAll()
+                }
+            }
+        }
+
+        private func closeIfNeeded() {
+            if !self.isClosed.load(), self.leased == 0, self.availableConnections.isEmpty, self.waiters.isEmpty {
+                self.isClosed.store(true)
+                self.parentPool._connectionProviders[key] = nil
+            }
+        }
+
+        struct Waiter {
+            var promise: EventLoopPromise<Connection>
+            var preference: HTTPClient.EventLoopPreference
+        }
+    }
+
+    private class HTTP2ConnectionProvider {
+        init(group: EventLoopGroup, key: Key, configuration: HTTPClient.Configuration, channel: Channel, multiplexer: HTTP2StreamMultiplexer, parentPool: ConnectionPool) {
+            self.loopGroup = group
+            self.key = key
+            self.configuration = configuration
+            self.multiplexer = multiplexer
+            self.channel = channel
+            self.parentPool = parentPool
+        }
+
+        let loopGroup: EventLoopGroup
+        let key: Key
+        let multiplexer: HTTP2StreamMultiplexer
+        let channel: Channel
+        let configuration: HTTPClient.Configuration
+        let lock = Lock()
+        unowned let parentPool: ConnectionPool
+        var closeFuture: EventLoopFuture<Void> {
+            return self.channel.closeFuture
+        }
+
+        func getConnection(preference: HTTPClient.EventLoopPreference) -> EventLoopFuture<Connection> {
+            return self.lock.withLock {
+                let promise = self.loopGroup.next().makePromise(of: Channel.self)
+                self.multiplexer.createStreamChannel(promise: promise) { channel, streamID in
+                    channel.pipeline.addHandler(HTTP2ToHTTP1ClientCodec(streamID: streamID, httpProtocol: .https))
+                }
+                return promise.futureResult.map { channel in
+                    Connection(key: self.key, channel: channel, parentPool: self.parentPool)
+                }
+            }
+        }
+
+        func release(connection: Connection) {
+            _ = connection.channel.close()
+        }
+
+        func closeAllConnections() -> EventLoopFuture<Void> {
+            return self.lock.withLock {
+                // FIXME: Is this correct regarding the child channels?
+                // I don't know how HTTP 2 "substreams" must be handled
+                channel.close()
+            }
+        }
+    }
+
+    class ConnectionRetryManager {
```

Comment:
This can be ignored for now and I'll remove it in an ulterior commit, I prefer to first fix the rest of the issues before coming back to the retry strategy implementation.